examples/hello_world/lib/e2e.dart

```diff
@@ -0,0 +1,183 @@
+            response = <String, String>{'message': 'pass'};
+            waitingForDriverScreenshot.complete(Future.value(true));
+            // Reset the pipe so if other tests wants to take a screenshot they can.
+            // If no tests want a screenshot the tearDownAll will complete
+            // the screenshotPipe completer.
+            screenshotPipe = Completer<String>();
+            break;
+        }
+        return <String, dynamic>{
+          'isError': false,
+          'response': response,
+        };
+      } else {
+        final String needAScreenshotFlag = await screenshotPipe.future;
+        if (needAScreenshotFlag.isNotEmpty) {
+          response = <String, String>{
+            'message': 'screenshot_$needAScreenshotFlag',
+          };
+          return <String, dynamic>{
+            'isError': false,
+            'response': response,
+          };
+        } else {
+          switch (command) {
+            case 'request_data':
+              final bool allTestsPassedFlag = await allTestsPassed.future;
+              response = <String, String>{
+                'message': allTestsPassedFlag ? 'pass' : 'fail',
+              };
+              break;
+            case 'get_health':
+              response = <String, String>{'status': 'ok'};
+              break;
+          }
+          return <String, dynamic>{
+            'isError': false,
+            'response': response,
+          };
+        }
+      }
+    }
+
+    if (kIsWeb) {
+      print('****************** kisweb');
+      registerWebServiceExtension(callback);
+      print('***************** web service extention registered');
+    }
+
+    registerServiceExtension(name: 'driver', callback: callback);
+  }
+
+  @override
+  Future<void> runTest(Future<void> testBody(), VoidCallback invariantTester,
+      {String description = '', Duration timeout}) async {
+    // TODO(jackson): Report the results individually instead of all at once
+    // See https://github.com/flutter/flutter/issues/38985
+    final TestExceptionReporter valueBeforeTest = reportTestException;
+    reportTestException =
+        (FlutterErrorDetails details, String testDescription) {
+      _results[description] = 'failed';
+      allTestsPassed.complete(false);
+      valueBeforeTest(details, testDescription);
+    };
+    await super.runTest(testBody, invariantTester,
+        description: description, timeout: timeout);
+    _results[description] ??= 'success';
+  }
+
+  @override
+  void reportExceptionNoticed(FlutterErrorDetails exception) {
+    print('exception $exception');
+  }
+
+  /// For flutter web tests we want to use the real messages instead of
+  /// test messages.
+  /// See [TestDefaultBinaryMessenger].
+  /// See [BinaryMessenger].
+  @override
+  BinaryMessenger createBinaryMessenger() {
+    if (kIsWeb) {
+      print('****************** kisweb return real binary messenger');
+      return super.createOriginalBinaryMessenger();
```

Comment:
The `TestDefaultBinaryMessenger` exposes the wrapped `BinaryMessenger` as a public `delegate` property. Can we take `super.createBinaryMessenger()`, assert that it's a `TestDefaultBinaryMessenger`, and read `delegate` so that we don't need to change the `TestWidgetsFlutterBinding` API?

It's also unclear to me why we need the real binary messenger here. It seems like all `TestDefaultBinaryMessenger` is doing is wrapping the real messenger and providing a `platformMessagesFinished` API.

Reply:
To be clear, what I'm proposing is:

```
TestDefaultBinaryMessenger messenger = super.createBinaryMessenger();
return messenger.delegate;
```

Reply:
That said, it may be fine to return `TestDefaultBinaryMessenger` since all it is doing is proxying calls to the delegate and exposing a `platformMessagesFinished` API.